pysingfel/experiment/base.py

```diff
@@ -0,0 +1,107 @@
+import numpy as np
+import pysingfel.geometry as psg
+from pysingfel.reciprocal_detector import ReciprocalDetector
+
+
+class Experiment(object):
+    mesh_size = 151
+
+    def __init__(self, det, beam, particles):
+        self.det = det
+        self.beam = beam
+        self.n_particle_kinds = len(particles)
+
+        # Create mesh
+        highest_k_beam = self.beam.get_highest_wavenumber_beam()
+        recidet = ReciprocalDetector(self.det, highest_k_beam)
+        mesh, self.voxel_length = recidet.get_reciprocal_mesh(
+            voxel_number_1d=self.mesh_size)
+
+        # Create volumes
+        import pysingfel.gpu as pg
+        self.volumes = []
+        for particle in particles:
+            if particle is None:
+                self.volumes.append(np.zeros(mesh.shape[:-1], np.complex128))
+                continue
+            self.volumes.append(
+                pg.calculate_diffraction_pattern_gpu(mesh, particle, return_type='complex_field'))
+
+    def generate_image(self):
+        img_stack = self.generate_image_stack()
+        return self.det.assemble_image_stack(img_stack)
+
+    def generate_image_stack(self, return_photons=None,
+                             return_intensities=False,
+                             always_tuple=False):
+        """
+        Generate and return a snapshot of the experiment.
+
+        By default, return a photon snapshot.
+        That behavior can be changed by setting:
+          - return_photons
+          - return_intensities
+        to True or False.
+
+        If more than one is requested, the function returns a tuple of
+        arrays instead of the array itself.
+        To return a tuple even if only one array is requested, set
+        always_tuple to True.
+        """
+        if return_photons is None and return_intensities is False:
+            return_photons = True
+
+        beam_spectrum = self.beam.generate_new_state()
+        sample_state = self.generate_new_sample_state()
+        intensities_stack = 0.
+        photons_stack = 0.
+
+        for spike in beam_spectrum:
+            recidet = ReciprocalDetector(self.det, spike)
+
+            group_complex_pattern = 0.
+            for i, particle_group in enumerate(sample_state):
+                group_complex_pattern += self._generate_group_complex_pattern(
```

Comment:
add comment here that we are summing up then converting to photons as opposed to converting to photons then summing up. Make a note that we may want to revisit the correctness of this procedure.

Reply:
comment added